scarlet/wavelet.py

```diff
@@ -0,0 +1,300 @@
+h = np.array([1. / 16, 1. / 4, 3. / 8, 1. / 4, 1. / 16])
+
+class Starlet(object):
+    """ A class used to create the Wavelet transform of a cube of images from the 'a trou' algorithm.
+
+        The transform is performed by convolving the image by a seed starlet: the transform of an all-zero
+        image with its central pixel set to one. This requires 2-fold padding of the image and an odd pad
+        shape. The fft of the seed starlet is cached so that it can be reused in the transform of other
+        images that have the same shape.
+    """
+    def __init__(self, image, lvl = None, starlet = None, direct = False):
+        """ Initialise the Starlet object
+
+        Paramters
+        image: Fourier object
+            image to transform
+        lvl: int
+            number of starlet levels to use in the decomposition
+        starlet: array
+            Starlet transform of an array
+        direct: bool
+            if set to True, uses direct wavelet transform with the a trou algorithm.
+            if set to False, the transform is performed by convolving the image by the wavelet transform of a dirac.
+        """
+        # Shape for the starlet padding. It is also an fft fast shape.
+        self.seed = None
+        if starlet is None:
+            self._starlet_shape = get_starlet_shape(image.shape, lvl = lvl)
+        else:
+            self._starlet_shape = starlet.shape
+        # Image (as a cube)
+        if len(image.shape) == 2:
+            self._image = image[np.newaxis, :, :]
+        else:
+            self._image = image
+        # Original shape of the image
+        self._image_shape = self._image.shape
+        # Can be initialised by a starlet transform to perform inverse transforms
+        if starlet is None:
+            if direct == True:
+                self._starlet = self.direct_transform()
+            else:
+                self._starlet = self.transform()
+        else:
+            if len(np.shape(starlet)) == 3:
+                self._starlet = starlet[np.newaxis, :, :, :]
+            else:
+                self._starlet = starlet
+
+        if self.seed is None:
+            self.seed = mk_starlet(self._starlet_shape)
+        self._norm = np.sqrt(np.sum(self.seed ** 2, axis=(-2, -1)))
+
+    @property
+    def image(self):
+        """The real space image"""
+        return self._image
+
+    @property
+    def norm(self):
+        """The norm of the seed wavelet in each wavelet level (not in coarse wavelet)"""
+        return self._norm
+
+
+    @property
+    def starlet(self):
+        """The real space image"""
+        return fft._centered(self._starlet,
+                             [self._image_shape[0], self._starlet_shape[-3], *self._image_shape[-2:]])
+
+    @property
+    def shape(self):
+        """The shape of the real space image"""
+        return self._image.shape
+
+    @staticmethod
+    def from_starlet(starlet, shape = None, starlet_shape = None):
```

Comment:
I'm not sure how necessary this method is. It's only useful if you have some starlet coefficients lying around ... And as far as I can tell it's only every used the tests and in the docs. In both cases, it's classed to check for the accuracy of the starlet representation of the original image, i.e. on the inverse transform of the class instance itself, so the coefficients don't need to be provided. I think this would be better done by redefining `@property def image` to return that representation instead of the original image.

Reply:
This method is here to create Starlet objects from a set of coefficients even when no image exist for instance after thresholding coefficients. I want to keep this, because I see wavelets better used as part of a constraint more than as a model. This would require the explicit transform and inverse. Also, this is consistent with the way Fourier transform and inverse is implemented.

Reply:
But in this case, the class doesn't behave properly. Say, you have transform coefficients and create an instance with this static method: `starlet = Starlet.from_starlet(coeffs)`. What is `starlet.image` or `starlet.norm`?

Reply:
Btw, here you can see again the problem I have with calling the transform coefficients `starlet`. It mixes up instances of the class `Starlet` with numpy arrays of the starlet coefficients.

Reply:
I suggest re-arranging a few items here:
* call this method `from_coefficients(coefficients, shape=None, starlet_shape=None)`
* don't keep reference to original image
* move the generation of an image from coefficients to `@property def image`.

The logic is the following. You normally instantiate the class with an image, but from then on the image itself is not needed any longer. By maintaining a reference you prevent the original from going out of scope, which is just waste of memory.

If you want to instantiate from a set of coefficients, use this method. In both cases, the image can/should be made only if needed, which would make this method faster as well.

Reply:
So what you want this to do is:
* Remove the 'starlet' argument from __init__
* compute the inverse in `from_coefficients`
* Declare a new Starlet object from the inverted image. 
Is that right? 
If so, no! Because the last bullet will call the starlet transform. But you don't need it because your coefficients already exist. This is supposed to be used when no image exist, only a set of coefficients!

Reply:
No, not right.
* You can keep `starlet` (or better `coefficients`) in `init`, that just saves computing the transformation if you already have it.
* Simple set the coefficents and shapes, etc in `from_coefficients`. Don't create the inverse.
* Create the inverse only when requested in `@property image`

My main point is, the `Starlet` class does not need the image array, it's fundamentally defined by the coefficients.

Reply:
For this to work, `init` needs to have optional `image` and `coefficient`, but one of them needs to be set. If you want to make them mutually exclusive or have coefficient takes preference is a matter of choice.

Reply:
I agree with your last point. With my suggestion you can get rid of the staticmethod (which probably should be a classmethod anyway), and simply do `Starlet(coefficients=coeffs).image`

Reply:
So optional image argument and then from_coefficients does not really need to exist anymore. __init__ will run the inverse transform?

Reply:
Yes, but why would init run the inverse transform? You seem to want to cache the inverse instead of creating it on-the-fly. Why is that?

Reply:
Because I don't see a reason to declare a Starlet object with a set of coefficients if it's not to get the inverse.

Reply:
Yeah, I get the same impression. For one, your code, also needs to call the `image` property explicitly, like here: https://github.com/pmelchior/scarlet/pull/166/files#diff-0bef36ec53a572489b9a02f7bc87f501R37, so in terms of convenience your current approach is as good as mine. But yours needs to create the image before it can even create the class and it keeps a reference to `image`. The first is not necessary (because coefficients are all you need) and the second is inefficient memory-wise.  

My suggestion avoid both problem by creating `image` on-the-fly, which has the extra advantage of being automatically consistent with the coefficients. This way you e.g. could modify the coefficient without having the generate a new `Starlet` instance solely for the purpose of getting a new inverse.